src/crypto/ocb_openssl.cc

```diff
@@ -1,1206 +1,162 @@
+  // the EVP_CIPHER_CTX wants to know if the operation is encryption
+  // or decryption. For now, arbitrary pick encryption. This will get
+  // overriden in the `ae_encrypt` and `ae_decrypt` methods when the
+  // key and nonce (IV) are loaded.
+  if (EVP_EncryptInit_ex(ctx->ocb_ctx, EVP_aes_128_ocb(),
+			 NULL, NULL, NULL) != 1) {
+    return -11;
+  }
+  // Attempt to set the nonce length. If it fails, the length must not
+  // be supported. However, that should have been handled by the
+  // pre-condition check above.
+  if (EVP_CIPHER_CTX_ctrl(ctx->ocb_ctx, EVP_CTRL_AEAD_SET_IVLEN,
+			  nonce_len, NULL)!= 1) {
+    return -12;
+  }
+  // Attempt to set the tag length. A NULL tag length means that
+  // EVP_CRTL_AEAD_SET_TAG is only being used to set the length
+  if (EVP_CIPHER_CTX_ctrl(ctx->ocb_ctx, EVP_CTRL_AEAD_SET_TAG,
+			  tag_len, NULL) != 1) {
+    return -13;
+  }
+  ctx->key = (unsigned char *)key_param;
+  ctx->tag_len = tag_len;
+  return AE_SUCCESS;
 }
 
-/* ----------------------------------------------------------------------- */
-
-static block gen_offset_from_nonce(ae_ctx *ctx, const void *nonce)
-{
-	const union { unsigned x; unsigned char endian; } little = { 1 };
-	union { uint32_t u32[4]; uint8_t u8[16]; block bl; } tmp;
-	unsigned idx;
-
-	/* Replace cached nonce Top if needed */
-	tmp.u32[0] = (little.endian?0x01000000:0x00000001);
-	tmp.u32[1] = ((uint32_t *)nonce)[0];
-	tmp.u32[2] = ((uint32_t *)nonce)[1];
-	tmp.u32[3] = ((uint32_t *)nonce)[2];
-	idx = (unsigned)(tmp.u8[15] & 0x3f);   /* Get low 6 bits of nonce  */
-	tmp.u8[15] = tmp.u8[15] & 0xc0;        /* Zero low 6 bits of nonce */
-	if ( unequal_blocks(tmp.bl,ctx->cached_Top) )   { /* Cached?       */
-		ctx->cached_Top = tmp.bl;          /* Update cache, KtopStr    */
-		AES_encrypt(tmp.u8, (unsigned char *)&ctx->KtopStr, &ctx->encrypt_key);
-		if (little.endian) {               /* Make Register Correct    */
-			ctx->KtopStr[0] = bswap64(ctx->KtopStr[0]);
-			ctx->KtopStr[1] = bswap64(ctx->KtopStr[1]);
-		}
-		ctx->KtopStr[2] = ctx->KtopStr[0] ^
-						 (ctx->KtopStr[0] << 8) ^ (ctx->KtopStr[1] >> 56);
-	}
-	return gen_offset(ctx->KtopStr, idx);
+int ae_encrypt(ae_ctx *ctx, const void *nonce_ptr, const void *pt_ptr,
+	       int pt_len, const void *ad_ptr, int ad_len, void *ct_ptr,
+	       void *tag, int final) {
+  int len = 0;
+  int ciphertext_len = 0;
+  const unsigned char *nonce = (const unsigned char *)nonce_ptr;
+  const unsigned char *pt = (const unsigned char *)pt_ptr;
+  const unsigned char* ad = (const unsigned char *)ad_ptr;
+  unsigned char* ct = (unsigned char *)ct_ptr;
+  // Set the key and nonce (IV) to the existing context if a nonce was
+  // provided. It will be omitted if `final` is set to AE_PENDING and
+  // "streaming" operation is in use.
+  if (nonce != NULL &&
+      EVP_EncryptInit_ex(ctx->ocb_ctx, NULL, NULL, ctx->key, nonce) != 1) {
+    return -10;
+  }
+  if (ad != NULL && ad_len > 0 &&
```

Comment:
If `ad_len > 0`, is `ad == NULL` allowed? I usually interpret ptr/len APIs to implicitly forbid this since (ptr, len) needs to be a valid span. A little weird that (NULL, 42) is implicitly empty.

On the flip side, be aware that EVP_CIPHER_CTX, particularly when used with AEADs, has lots of stupid behavior around NULL. If you pass `in == NULL` to EncryptUpdate, it'll get confused and think you're calling EncryptFinal. If you pass `out == NULL`, it'll do the AAD thing.